src/classes/OPP_SendAcknowledgmentBTN_TEST.cls

```diff
@@ -35,43 +35,89 @@
-        insert testocr;
-        
-        list<Opportunity> listOpp = new list<Opportunity>{testoppty};
+    static testmethod void testSendAcknowledgment() {        
+        list<Opportunity> listOpp = setUpOpportunities(1);
         ApexPages.StandardSetController sc = new ApexPages.StandardSetController(listOpp);
         sc.setSelected(listOpp);
-        OPP_SendAcknowledgmentBTN_CTRL btnCtrl = new OPP_SendAcknowledgmentBTN_CTRL(sc);
+        OPP_SendAcknowledgmentBTN_CTRL ctrl = new OPP_SendAcknowledgmentBTN_CTRL(sc);
         
-        test.startTest();
-        btnCtrl.SendAcknowledgment();
-        test.stopTest();
+        Test.startTest();
+        ctrl.SendAcknowledgment();
+        Test.stopTest();
         
         listOpp = [select Id, name, Acknowledgment_Status__c, Acknowledgment_Date__c from Opportunity];
         system.assertEquals(1, listOpp.size());
+
         // we don't know if the workflow is active or not.
         system.assert(listOpp[0].Acknowledgment_Status__c == label.sendAcknowledgmentFailedStatus ||
             listOpp[0].Acknowledgment_Status__c == label.sendAcknowledgmentSucceeded);            
-        system.assertNotEquals(null, btnCtrl.listOppNotSent);
-        system.assertNotEquals(btnCtrl.cSent, btnCtrl.cNotSent);
+        system.assertNotEquals(null, ctrl.listOppNotSent);
+        system.assertNotEquals(ctrl.cSent, ctrl.cNotSent);   
+    }
+
+    /*********************************************************************************************************
+    @description 
+        Test SendAcknowledgment() rollback when an exception occurs 
+    verify:
+        The transaction is completely rolled back and there is no partial commit
+    **********************************************************************************************************/ 
+    private static testMethod void testSendAcknowledgmentRollsBackAllDmlsOnError() { 
+        Test.setCurrentPage(Page.OPP_SendAcknowledgmentBTN);
+
+        List<Opportunity> opps = setUpOpportunities(2);
+        //fail an update
+        opps[0].Id = Opportunity.sObjectType.getDescribe().getKeyPrefix() + '000000000001AAA';
+
+        ApexPages.StandardSetController sc = new ApexPages.StandardSetController(opps);
+        sc.setSelected(opps);
+        OPP_SendAcknowledgmentBTN_CTRL ctrl = new OPP_SendAcknowledgmentBTN_CTRL(sc);
+        
+        Test.startTest();
+        ctrl.SendAcknowledgment();
+        Test.stopTest();
+
+        UTIL_UnitTestData_TEST.assertPageHasError('invalid cross reference');
+
+        for (Opportunity opp : [SELECT Acknowledgment_Status__c FROM Opportunity]) {
+            System.assertEquals(TO_BE_ACKNOWLEDGED_STATUS, opp.Acknowledgment_Status__c, 'The Acknowledgment Status should be unchanged');
+        }     
+    }
+
+    // Helpers
+    ////////////
+
+    private static List<Opportunity> setUpOpportunities(Integer oppCount) {
```

Comment:
even though this is a private helper method in a test class, it would still be nice to still see it with its ApexDoc function description.

Reply:
This has been overlooked, it will be fixed. Thanks.